### PR TITLE
feat(routing): build gRPC Client Interface to initiate communication with other gRPC services

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Feature Description:
 Building a Grpc Client for communicating with the Dynamo server, once the dynamic routing feature has been enabled.

The dynamic routing feature  would be a compile time as well as runtime feature flag ,
 for providing the flexibility to enable and disable the dynamic routing during deployment

Possible Implementation:
Once dynamic routing has been enabled , we will store the default config in the algorithm column in routing table as the following jsonb:

```
{
    "id": "merchant_id", 
    "params": "card", 
    "labels": ["stripe","adyen"],
    "config": {
        "min_aggregates_size": 2,
        "default_success_rate": 100.0
    }
}
```

gRPC  client will be integrated as a struct of  GrpcClientInterface within the struct  SessionState, 
```
struct SessionState
{
	grpc:  GrpcClientInterface
}
```

The gRPcClientInterface  would be a struct which would have Dynamic Routing as its field and similarly any other future gRPC Client would be one of its fields

```
struct gRPcClientInterface
{
pub dynamic_routing: DynamicRouting
}
```

The DynamicRouting struct would have all the fields that would be there as its services channel

```
struct DynamicRouting
{
pub success_based_client : SuccessBasedClient<Channel>
}

```

The SuccessBasedClient would be a trait would have following  functions: 
calculate_the_success_rate 
update_the_success_rates

```
triat SuccessBasedClient 
{
	type CalCulatedRespons;
	type UpadationResponse;

	async fn calculate_success_rate(self,label_input : Vec<T>) -> CalCulatedRespons

	async fn update_success_rate(self,calculated_success_rate:CalCulatedRespons) -> UpadationResponse
}
```

The SuccessBasedClient would be implemented on client that would be a SuccessBasedClient<Channel>

```
impl SuccessBasedClient for SuccessBasedClient<Channel>
{
	async fn calculate_success_rate(self,label_input : T) ->     CalSuccessRateResponse
	{
		self.fetch_success_rate(request)
	}

	async fn update_success_rate(self,calculated_success_rate :  CalSuccessRateResponse) -> UpdateSuccessRateWindowResponse
  	{

   		self.update_success_rate_window(request)
   	}

}
```

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
- This cannot be tested on sbx, yet

I have tested it locally
- Started the gRPC Server
- Established the connection with the Client <-> Server
- Sent Request the gRPC Server
Screenshot 2024-09-17 at 7 29 17 PM

- Response from the server
Screenshot 2024-09-17 at 7 30 27 PM

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/external_services/Cargo.toml [hunk 1]
```diff
@@ -31,14 +32,23 @@ hyper-proxy = "0.9.1"
 once_cell = "1.19.0"
 serde = { version = "1.0.197", features = ["derive"] }
 thiserror = "1.0.58"
-tokio = "1.37.0"
 vaultrs = { version = "0.7.2", optional = true }
+prost = "0.13"
+tokio = { version = "1.37.0", features = ["macros", "rt-multi-thread"] }
+tonic = "0.12.2"
+tonic-reflection = "0.12.2"
+tonic-types = "0.12.2"
```
Comment (Chethan-rao) [inline crates/external_services/Cargo.toml:40]: Can we make these dependencies optional, only pulling them in when the relevant gRPC features are enabled

File: crates/external_services/src/grpc_client.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+pub mod dynamic_routing {
+    tonic::include_proto!("success_rate");
+}
+use serde;
+
+pub type DRResult<T> = CustomResult<T, DRError>;
+
+// The trait Success Based Dynamic Routing would have all the functions required to support the calculation and updation window
+#[async_trait::async_trait]
+pub trait SuccessBasedDynamicRouting: dyn_clone::DynClone + Send + Sync {
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client.rs:33]: Can we have a proper module hierarchy here? move these DR related types and implementation under grpc_client/dynamic_routing module

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

File: config/development.toml [hunk 3]
```diff
@@ -731,5 +731,11 @@ encryption_key = "A8EF32E029BC3342E54BF2E172A4D7AA43E8EF9D2C3A624A9F04E2EF79DC69
 [locker_based_open_banking_connectors]
 connector_list = ""
 
+[grpc_client]
```
Comment (NishantJoshi00) [inline config/development.toml:734]: This might note be rquired

File: config/development.toml [hunk 4]
```diff
@@ -731,5 +731,11 @@ encryption_key = "A8EF32E029BC3342E54BF2E172A4D7AA43E8EF9D2C3A624A9F04E2EF79DC69
...
+[grpc_client.dynamic_routing_client]
+host = "localhost"
+port = 7000
+enabled = false
```
Comment (NishantJoshi00) [inline config/development.toml:738]: If you say enabled is `false` then shouldn't all the fields be optional? Maybe something like a tagged enum?

File: crates/external_services/src/grpc_client.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    UpdateSuccessRateWindowConfig, UpdateSuccessRateWindowRequest, UpdateSuccessRateWindowResponse,
+};
+use error_stack::ResultExt;
+use hyperswitch_interfaces::api::api_models::routing::{
+    CurrentBlockThreshold, DynamicRoutingConfig, DynamicRoutingConfigBody, RoutableConnectorChoice,
+    RoutableConnectorChoiceWithStatus,
+};
+use router_env::logger;
+use tonic::transport::Channel;
+#[allow(
```
Comment (NishantJoshi00) [inline crates/external_services/src/grpc_client.rs:18]: I don't feel right if we evade `clippy` this way, what do you guys think?

cc: @SanchithHegde @Chethan-rao

File: crates/external_services/src/grpc_client.rs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+pub struct DynamicRoutingClientConfig {
+    pub host: String,
+    pub port: u16,
+    pub enabled: bool,
+}
+
+impl GrpcClientSettings {
+    /// # Panics
+    ///
+    /// This function will panic if it fails to establish a connection with the gRPC server.
```
Comment (NishantJoshi00) [inline crates/external_services/src/grpc_client.rs:71]:
```suggestion
    /// This function will panic if it fails to establish a connection with the gRPC server.
    /// This function will be called at service startup.
```

File: crates/external_services/src/grpc_client.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            .params
+            .map(|vec| {
+                if !vec.is_empty() {
+                    vec.into_iter()
+                        .map(|param| param.to_string())
+                        .collect::<Vec<String>>()
+                        .join(":")
+                } else {
+                    String::default()
+                }
```
Comment (NishantJoshi00) [inline crates/external_services/src/grpc_client.rs:131]:
```suggestion
                input.into_iter().fold(String::new(), |mut acc, cur| {
                        if !acc.is_empty() {
                            acc.push(':');
                        }
                        acc.push_str(cur);
                        acc
                });
```

This can solve for the following:
1. Less memory allocations
2. The outer check for empty vector
3. Default behavior is implicit

File: crates/external_services/src/grpc_client.rs [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            .update_success_rate_window(request)
+            .await
+            .change_context(DRError::GrpcServerResponseFailure)?
+            .into_inner();
+
+        Ok(response)
+    }
+}
+
+impl ForeignFrom<CurrentBlockThreshold> for DynamicCurrentThreshold {
```
Comment (NishantJoshi00) [inline crates/external_services/src/grpc_client.rs:211]: Why can't we have a traditional `From` here?

File: crates/external_services/src/lib.rs [hunk 9]
```diff
@@ -14,6 +14,8 @@ pub mod hashicorp_vault;
 
 pub mod no_encryption;
 
+#[allow(missing_docs)]
```
Comment (NishantJoshi00) [inline crates/external_services/src/lib.rs:17]: can we add documentation for this?

File: crates/router/src/core/payments.rs [hunk 10]
```diff
@@ -34,6 +33,7 @@ use common_utils::{
 use diesel_models::{ephemeral_key, fraud_check::FraudCheck};
 use error_stack::{report, ResultExt};
 use events::EventInfo;
+// use external_services::grpc_client::SuccessBasedDynamicRouting;
```
Comment (NishantJoshi00) [inline crates/router/src/core/payments.rs:36]: Is this required?

File: crates/router/src/routes/api_keys.rs [hunk 11]
```diff
@@ -166,7 +166,7 @@ pub async fn api_key_update(
     payload.key_id = key_id;
     payload.merchant_id.clone_from(&merchant_id);
 
-    api::server_wrap(
+    Box::pin(api::server_wrap(
```
Comment (NishantJoshi00) [inline crates/router/src/routes/api_keys.rs:169]: Why is this required?

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/router/src/routes/api_keys.rs (hunk 11 above)
Comment (NishantJoshi00) [reply crates/router/src/routes/api_keys.rs:169]: `clippy:large_futures`?

Comment (Aprabhat19) [reply crates/router/src/routes/api_keys.rs:169]: Was addressing clippy lints, for `large_futures`

File: crates/external_services/src/lib.rs (hunk 9 above)
Comment (Aprabhat19) [reply crates/external_services/src/lib.rs:17]: We cannot add docs to the external file

File: crates/router/src/core/payments.rs (hunk 10 above)
Comment (Aprabhat19) [reply crates/router/src/core/payments.rs:36]: Nope , my bad missed it during testing. Will remove it

File: config/development.toml (hunk 3 above)
Comment (Aprabhat19) [reply config/development.toml:734]: This is so we can have all the hosts and ports for all the clients under one `grpc_client` struct

File: crates/external_services/src/grpc_client.rs (hunk 8 above)
Comment (Aprabhat19) [reply crates/external_services/src/grpc_client.rs]: Cause of the orphan rule, since ` CurrentBlockTreshold` is in another crate

File: crates/external_services/Cargo.toml [hunk 12]
```diff
@@ -31,14 +32,23 @@ hyper-proxy = "0.9.1"
 once_cell = "1.19.0"
 serde = { version = "1.0.197", features = ["derive"] }
 thiserror = "1.0.58"
-tokio = "1.37.0"
 vaultrs = { version = "0.7.2", optional = true }
+prost = { version = "0.13", optional = true }
+tokio = { version = "1.37.0", features = ["macros", "rt-multi-thread"] }
+tonic = { version = "0.12.2", optional = true }
+tonic-reflection = { version = "0.12.2", optional = true }
+tonic-types ={ version = "0.12.2", optional = true }
```
Comment (Chethan-rao) [inline crates/external_services/Cargo.toml:40]:
```suggestion
tonic-types = { version = "0.12.2", optional = true }
```

File: crates/external_services/src/grpc_client.rs [hunk 13]
```diff
@@ -0,0 +1,46 @@
+#[cfg(feature = "dynamic_routing")]
+pub mod dynamic_routing;
+use std::fmt::Debug;
+
+use router_env::logger;
+use serde;
+
+#[cfg(feature = "dynamic_routing")]
+use crate::grpc_client::dynamic_routing::{DynamicRoutingClientConfig, RoutingStrategy};
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client.rs:9]:
```suggestion
use dynamic_routing::{DynamicRoutingClientConfig, RoutingStrategy};
```

File: crates/external_services/src/grpc_client.rs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    /// This function will be called at service startup.
+    #[allow(clippy::expect_used)]
+    pub async fn get_grpc_client_interface(&self) -> GrpcClients {
+        #[cfg(feature = "dynamic_routing")]
+        let dynamic_routing_connection = self
+            .dynamic_routing_client
+            .clone()
+            .get_dynamic_routing_connection()
+            .await
+            .expect("Failed to establish a connection with the gRPC Server");
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client.rs:37]:
```suggestion
            .expect("Failed to establish a connection with the dynamic routing gRPC Server");
```

File: crates/router/src/routes/app.rs [hunk 15]
```diff
@@ -105,6 +105,7 @@ pub struct SessionState {
     pub tenant: Tenant,
     #[cfg(feature = "olap")]
     pub opensearch_client: Arc<OpenSearchClient>,
+    pub grpc_client: GrpcClients,
```
Comment (Chethan-rao) [inline crates/router/src/routes/app.rs:108]: Should we keep these clients behind `Arc`?

File: migrations/2024-09-05-160455_add_new_col_is_dynamic_routing_algorithm_in_business_profile/up.sql [hunk 16]
```diff
@@ -1,3 +1,5 @@
 -- Your SQL goes here
-ALTER TABLE business_profile
-ADD COLUMN dynamic_routing_algorithm JSON DEFAULT NULL;
+ALTER TABLE
+    business_profile
+ADD
+    COLUMN dynamic_routing_algorithm JSON DEFAULT NULL;
```
Comment (Chethan-rao) [inline migrations/2024-09-05-160455_add_new_col_is_dynamic_routing_algorithm_in_business_profile/up.sql:5]: Could u revert this formatting. Otherwise CI adds unnecessary DB changes label

File: crates/hyperswitch_interfaces/src/api.rs [hunk 17]
```diff
@@ -16,7 +16,7 @@ pub mod payouts;
 pub mod payouts_v2;
 pub mod refunds;
 pub mod refunds_v2;
-
+pub use api_models;
```
Comment (Chethan-rao) [inline crates/hyperswitch_interfaces/src/api.rs:19]: why are you re-exporting entire crate here?

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+
+        let labels = label_input
+            .into_iter()
+            .map(|conn_choice| conn_choice.to_string())
+            .collect::<Vec<_>>();
+
+        let config = dynamic_routing_config.config.map(ForeignFrom::foreign_from);
+
+        let request = tonic::Request::new(CalSuccessRateRequest {
+            id: id.get_string_repr().to_owned(),
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:119]: Should we also have `tenant_id` prefix to id here so that all the dynamic routing id's related to one tenant is logically grouped

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+            params,
+            labels,
+            config,
+        });
+
+        let mut client = self.clone();
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:125]: why is this clone required? Can we instead directly receive `&mut self` in the function?

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+
+        let response = client
+            .fetch_success_rate(request)
+            .await
+            .change_context(DRError::GrpcServerResponseFailure)?
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:130]: `GrpcServerResponseFailure` alone cannot be reason why this call would have failed right? It might also be genuine error from `Dynamo` too.

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+                    vec.into_iter()
+                        .map(|param| param.to_string())
+                        .collect::<Vec<_>>()
+                        .join(":")
+                } else {
+                    String::default()
+                }
+            })
+            .get_required_value("Vector of params")
+            .change_context(DRError::ClientBuildingFailed)?;
```
Comment (Chethan-rao) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:109]: Why is this mapped to client build failure?

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

File: crates/external_services/src/grpc_client/dynamic_routing.rs (hunk 19 above)
Comment (Aprabhat19) [reply crates/external_services/src/grpc_client/dynamic_routing.rs:125]: Nope we cannot do that

File: crates/external_services/src/grpc_client/dynamic_routing.rs (hunk 21 above)
Comment (Aprabhat19) [reply crates/external_services/src/grpc_client/dynamic_routing.rs:109]: Cause the params needed to build the client was empty.

Comment (Chethan-rao) [reply crates/external_services/src/grpc_client/dynamic_routing.rs:117]: client is already built during startup right. This can happen when the params itself is not found in db for a profile as it is Option. I think we should say necessary params not found for making request

File: crates/external_services/build.rs [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Get the directory of the current crate
+    let crate_dir = env::var("CARGO_MANIFEST_DIR")?;
+    let proto_file = PathBuf::from(crate_dir)
+        .join("..")
+        .join("..")
+        .join("proto")
+        .join("success_rate.proto");
+    // Compile the .proto file
+    tonic_build::compile_protos(proto_file).expect("Failed to compile protos ");
```
Comment (prajjwalkumar17) [inline crates/external_services/build.rs:13]:
```suggestion
    tonic_build::compile_protos(proto_file).expect("Failed to compile sucess_rate.proto file");
```

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+)]
+pub mod success_rate {
+    tonic::include_proto!("success_rate");
+}
+
+pub type DRResult<T> = CustomResult<T, DRError>;
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum DRError {
+    #[error("Error buling the gRPC Client for communication")]
```
Comment (prajjwalkumar17) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:35]:
```suggestion
    #[error("Error building dynamic routing gRPC client for communication")]
```

File: crates/external_services/src/grpc_client.rs [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+
+        logger::info!("Connection established with Grpc Server");
```
Comment (prajjwalkumar17) [inline crates/external_services/src/grpc_client.rs:42]:
```suggestion
        logger::info!("Connection established with gRPC Server");
```
can we have uniform naming of gRPC varibales and occurences in logs? 
If not in structs or fields, possibly in logs.
As it is right now mixup of Grpc and gRPC.

File: crates/external_services/src/grpc_client/dynamic_routing.rs (hunk 21 above)
Comment (Aprabhat19) [reply crates/external_services/src/grpc_client/dynamic_routing.rs:117]: Ok

**Review by prajjwalkumar17 — CHANGES_REQUESTED** [review]

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+/// The trait Success Based Dynamic Routing would have the functions required to support the calculation and updation window
+#[async_trait::async_trait]
+pub trait SuccessBasedDynamicRouting: dyn_clone::DynClone + Send + Sync {
+    /// To calculate the success rate for the list of chosen connectors
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
```
Comment (prajjwalkumar17) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:85]:
```suggestion
        success_based_routing_config: SuccessBasedRoutingConfig,
```

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DynamicRoutingResult<CalSuccessRateResponse>;
+    /// To update the success rate with the given label
+    async fn update_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
```
Comment (prajjwalkumar17) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:92]: here as well

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+        response: Vec<RoutableConnectorChoiceWithStatus>,
+    ) -> DynamicRoutingResult<UpdateSuccessRateWindowResponse>;
+}
+
+#[async_trait::async_trait]
+impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
```
Comment (prajjwalkumar17) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:102]: same here

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+            .change_context(DynamicRoutingError::GrpcServerResponseFailure)?
+            .into_inner();
+
+        Ok(response)
+    }
+
+    async fn update_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
```
Comment (prajjwalkumar17) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:147]:
```suggestion
        success_based_routing_config: SuccessBasedRoutingConfig,
```
Let's try to change this in every possible places so that the model will be prepared beforehand, for other variants of Dynamic Routing.

**Review by prajjwalkumar17 — CHANGES_REQUESTED** [review]

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 29]
```diff
@@ -0,0 +1,265 @@
...
+    },
+    #[default]
+    /// If the dynamic routing client config has been disabled
+    Disabled,
+}
+
+impl DynamicRoutingClientConfig {
+    /// establish connection with the server
+    pub async fn get_dynamic_routing_connection(
+        self,
```
Comment (NishantJoshi00) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:69]: nit:
```suggestion
        &self,
```

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

**Review by NishantJoshi00 — DISMISSED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]
Dashboard changes look fine.

**Review by prajjwalkumar17 — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

File: crates/external_services/build.rs [hunk 30]
```diff
@@ -0,0 +1,15 @@
...
+
+#[allow(clippy::expect_used)]
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Get the directory of the current crate
+    let crate_dir = env::var("CARGO_MANIFEST_DIR")?;
+    let proto_file = PathBuf::from(crate_dir)
+        .join("..")
+        .join("..")
+        .join("proto")
+        .join("success_rate.proto");
```
Comment (SanchithHegde) [inline crates/external_services/build.rs:11]: Use `router_env::workspace_path()` preferably.

Can take up in a separate PR.

File: crates/external_services/src/grpc_client.rs [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+    /// This function will be called at service startup.
+    #[allow(clippy::expect_used)]
+    pub async fn get_grpc_client_interface(&self) -> Arc<GrpcClients> {
+        #[cfg(feature = "dynamic_routing")]
+        let dynamic_routing_connection = self
+            .dynamic_routing_client
+            .clone()
+            .get_dynamic_routing_connection()
+            .await
+            .expect("Failed to establish a connection with the Dynamic Routing Server");
```
Comment (SanchithHegde) [inline crates/external_services/src/grpc_client.rs:39]: But do we necessarily _need_ to panic if we fail to establish a connection?

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 32]
```diff
@@ -0,0 +1,265 @@
...
+    ) -> Result<RoutingStrategy, Box<dyn std::error::Error>> {
+        let success_rate_client = match self {
+            Self::Enabled { host, port } => {
+                let uri = format!("http://{}:{}", host, port);
```
Comment (SanchithHegde) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:73]: Do we ever plan to have TLS support? Asking since the scheme is hardcoded as `http`.

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 33]
```diff
@@ -0,0 +1,265 @@
...
+    async fn calculate_success_rate(
+        &self,
+        id: String,
+        success_rate_based_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DynamicRoutingResult<CalSuccessRateResponse> {
+        let params = success_rate_based_config
+            .params
+            .map(|vec| {
+                vec.into_iter().fold(String::new(), |mut acc_vec, params| {
```
Comment (SanchithHegde) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:115]: `acc_vec` is a `String` here and not a `Vec`, isn't it? Can we consider using a different variable name?

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 34]
```diff
@@ -0,0 +1,265 @@
...
+            .map(|conn_choice| LabelWithStatus {
+                label: conn_choice.routable_connector_choice.to_string(),
+                status: conn_choice.status,
+            })
+            .collect();
+
+        let params = success_rate_based_config
+            .params
+            .map(|vec| {
+                vec.into_iter().fold(String::new(), |mut acc_vec, params| {
```
Comment (SanchithHegde) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:180]: Same here.

File: crates/external_services/src/grpc_client.rs (hunk 31 above)
Comment (Aprabhat19) [reply crates/external_services/src/grpc_client.rs:39]: By default we would have gRPC as disabled but once enabled and we would want a connection to be established with the server. But yes to panic the entire application or not we can discuss about it , logging can also be an option
cc : @NishantJoshi00  @Chethan-rao
